### ISSUE TITLE
Errors with symlinks and Google Drive

### ISSUE THREAD

**Issue by ruphy** [rclone/rclone#4950]
#### What is the problem you are having with rclone?

I am backing up some LucidLink storage (a cloud storage provider that mounts with its daemon as a fuse space) to Google Drive, and links don't get copied because of a "detected corruption". The issue seems similar to #4581, however it persists even in the last beta.

Interestingly, it seems to be a problem only with files stored on the Lucid mount, as I confirm that copying symlinks that exist on a traditional filesystem works just fine. Copying a directory with links from the Lucid mount to my homedir and retrying the copy from there makes everything work, so the bug definitely exists between the interaction of the two tools.

However, symlinks exist properly within the lucid virtual filesystem and are otherwise properly behaving and working in any other app I tried so far (I have been using them extensively since a while niw). I can't get my head around why rclone should fail.

This is the test layout I am trying to copy over:
```
~/Lucid 
❯ file testdir/*
testdir/link-dir:  symbolic link to test1
testdir/link-file: symbolic link to test2
testdir/test1:     directory
testdir/test2:     empty
```

#### What is your rclone version (output from `rclone version`)

`v1.54.0-beta.5087.3e986cdf5`

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Linux (Elementary OS - Ubuntu LTS)

####  Which cloud storage system are you using? (e.g. Google Drive)

Google Drive

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`❯ rclone copy -vv -Pl testdir gdrive:testdir `

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
~/Lucid
❯ rclone copy -vv -Pl testdir gdrive:testdir 
2021/01/22 19:15:39 DEBUG : rclone: Version "v1.54.0-beta.5087.3e986cdf5" starting with parameters ["rclone" "copy" "-vv" "-Pl" "testdir" "gdrive:testdir"]
2021/01/22 19:15:39 DEBUG : Creating backend with remote "testdir"
2021/01/22 19:15:39 DEBUG : Using config file from "/home/ruphy/.config/rclone/rclone.conf"
2021/01/22 19:15:39 DEBUG : fs cache: renaming cache item "testdir" to be canonical "/home/ruphy/Lucid/testdir"
2021/01/22 19:15:39 DEBUG : Creating backend with remote "gdrive:testdir"
2021/01/22 19:15:39 DEBUG : gdrive: Loaded invalid token from config file - ignoring
2021/01/22 19:15:39 DEBUG : gdrive: Saved new token in config file
2021-01-22 19:15:41 DEBUG : Google drive root 'testdir': Waiting for checks to finish
2021-01-22 19:15:41 DEBUG : Google drive root 'testdir': Waiting for transfers to finish
2021-01-22 19:15:44 DEBUG : test2: MD5 = d41d8cd98f00b204e9800998ecf8427e OK
2021-01-22 19:15:44 INFO  : test2: Copied (new)
2021-01-22 19:15:46 DEBUG : test1/empty: MD5 = d41d8cd98f00b204e9800998ecf8427e OK
2021-01-22 19:15:46 INFO  : test1/empty: Copied (new)
2021-01-22 19:15:46 ERROR : link-file.rclonelink: corrupted on transfer: sizes differ 0 vs 5
2021-01-22 19:15:46 INFO  : link-file.rclonelink: Removing failed copy
2021-01-22 19:15:46 ERROR : link-dir.rclonelink: corrupted on transfer: sizes differ 0 vs 5
2021-01-22 19:15:46 INFO  : link-dir.rclonelink: Removing failed copy
2021-01-22 19:15:47 ERROR : Attempt 1/3 failed with 2 errors and: corrupted on transfer: sizes differ 0 vs 5
2021-01-22 19:15:47 DEBUG : test2: Size and modification time the same (differ by -452.558µs, within tolerance 1ms)
2021-01-22 19:15:47 DEBUG : test2: Unchanged skipping
2021-01-22 19:15:48 DEBUG : test1/empty: Size and modification time the same (differ by -558.716µs, within tolerance 1ms)
2021-01-22 19:15:48 DEBUG : test1/empty: Unchanged skipping
2021-01-22 19:15:48 DEBUG : Google drive root 'testdir': Waiting for checks to finish
2021-01-22 19:15:48 DEBUG : Google drive root 'testdir': Waiting for transfers to finish
2021-01-22 19:15:49 ERROR : link-dir.rclonelink: corrupted on transfer: sizes differ 0 vs 5
2021-01-22 19:15:49 INFO  : link-dir.rclonelink: Removing failed copy
2021-01-22 19:15:49 ERROR : link-file.rclonelink: corrupted on transfer: sizes differ 0 vs 5
2021-01-22 19:15:49 INFO  : link-file.rclonelink: Removing failed copy
2021-01-22 19:15:50 ERROR : Attempt 2/3 failed with 2 errors and: corrupted on transfer: sizes differ 0 vs 5
2021-01-22 19:15:50 DEBUG : test2: Size and modification time the same (differ by -452.558µs, within tolerance 1ms)
2021-01-22 19:15:50 DEBUG : test2: Unchanged skipping
2021-01-22 19:15:51 DEBUG : test1/empty: Size and modification time the same (differ by -558.716µs, within tolerance 1ms)
2021-01-22 19:15:51 DEBUG : test1/empty: Unchanged skipping
2021-01-22 19:15:51 DEBUG : Google drive root 'testdir': Waiting for checks to finish
2021-01-22 19:15:51 DEBUG : Google drive root 'testdir': Waiting for transfers to finish
2021-01-22 19:15:52 ERROR : link-file.rclonelink: corrupted on transfer: sizes differ 0 vs 5
2021-01-22 19:15:52 INFO  : link-file.rclonelink: Removing failed copy
2021-01-22 19:15:52 ERROR : link-dir.rclonelink: corrupted on transfer: sizes differ 0 vs 5
2021-01-22 19:15:52 INFO  : link-dir.rclonelink: Removing failed copy
2021-01-22 19:15:53 ERROR : Attempt 3/3 failed with 2 errors and: corrupted on transfer: sizes differ 0 vs 5
Transferred:            30 / 30 Bytes, 100%, 2 Bytes/s, ETA 0s
Errors:                 2 (retrying may help)
Checks:                 4 / 4, 100%
Transferred:            2 / 2, 100%
Elapsed time:        14.0s
2021/01/22 19:15:53 INFO  : 
Transferred:            30 / 30 Bytes, 100%, 2 Bytes/s, ETA 0s
Errors:                 2 (retrying may help)
Checks:                 4 / 4, 100%
Transferred:            2 / 2, 100%
Elapsed time:        14.0s

2021/01/22 19:15:53 DEBUG : 14 go routines active
2021/01/22 19:15:53 Failed to copy with 2 errors: last error was: corrupted on transfer: sizes differ 0 vs 5
```

**Comment by ruphy** [comment]
I have taken a look at https://github.com/rclone/rclone/commit/cc17cbf03aded2aaf630687f6950c6a70b32c8b6 and it seems like we could be in a similar case here: a quick check with du report that links on this virtual filesystem are also zero in size, even if it's not a windows machine.

I wouldn't know how to elegantly and reliably detect we are in this configuration, so the only option I am imagining is to implement a flag such as "--zero-size-links" or the likes for users who face a similar issue. Or maybe something else I haven't imagined.

Is there any preferred solution?

**Comment by ncw** [comment]
> I wouldn't know how to elegantly and reliably detect we are in this configuration, so the only option I am imagining is to implement a flag such as "--zero-size-links" or the likes for users who face a similar issue. Or maybe something else I haven't imagined.

This seems like a reasonable option if you want to have a go?

What you'd do is create a new option like this one

https://github.com/rclone/rclone/blob/127f48e8ad3ef85dabfbdbe15ad9149a84a690e1/backend/local/local.go#L147-L157

Add it here

https://github.com/rclone/rclone/blob/127f48e8ad3ef85dabfbdbe15ad9149a84a690e1/backend/local/local.go#L180

And use it here

https://github.com/rclone/rclone/blob/127f48e8ad3ef85dabfbdbe15ad9149a84a690e1/backend/local/local.go#L1234-L1235

**Comment by ruphy** [comment]
I'll give it a go! Thanks for the great pointers!

**Comment by ruphy** [comment]
done :wink: and I confirm it fixes my issues

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54

Thanks @ruphy for fixing this :-)
